benchbuild/command.py

```diff
@@ -0,0 +1,482 @@
+        assert cmd_path.exists(), f"{str(cmd_path)} doesn't exist!"
+
+        cmd = local[str(cmd_path)]
+        cmd_w_args = cmd[self._args]
+        cmd_w_output = cmd_w_args
+        if self.output:
+            output_path = self.output.render(**kwargs)
+            output_params = [
+                arg.format(output=output_path) for arg in self._output_param
+            ]
+            cmd_w_output = cmd_w_args[output_params]
+        cmd_w_env = cmd_w_output.with_env(**self._env)
+
+        return cmd_w_env
+
+    def __repr__(self) -> str:
+        repr_str = f"path={self._path}"
+
+        if self._args:
+            repr_str += f" args={self._args}"
+        if self._env:
+            repr_str += f" env={self._env}"
+        if self._output:
+            repr_str += f" output={self._output}"
+        if self._output_param:
+            repr_str += f" output_param={self._output_param}"
+
+        return f"Command({repr_str})"
+
+    def __str__(self) -> str:
+        env_str = " ".join([f"{k}={str(v)}" for k, v in self._env.items()])
+        args_str = " ".join(self._args)
+
+        return f"{env_str} {self._path} {args_str}"
+
+
+class ProjectCommand:
+    """ProjectCommands associate a command to a benchbuild project.
+
+    A project command can wrap the given command with the assigned
+    runtime extension.
+    If the binary is located inside a subdirectory relative to one of the
+    project's sources, you can provide a path relative to it's local
+    directory.
+    A project command will always try to resolve any reference to a local
+    source directory in a command's path.
+
+    A call to a project command will drop the current configuration inside
+    the project's build directory and confine the run into the project's
+    build directory. The binary will be replaced with a wrapper that
+    calls the project's runtime_extension.
+    """
+
+    project: "benchbuild.project.Project"
+    command: Command
+
+    def __init__(
+        self, project: "benchbuild.project.Project", command: Command
+    ) -> None:
+        self.project = project
+        self.command = command
+
+    @property
+    def path(self) -> PathToken:
+        return self.command.path
+
+    def __call__(self, *args: tp.Any):
+        build_dir = self.project.builddir
+
+        CFG.store(Path(build_dir) / ".benchbuild.yml")
+        with local.cwd(build_dir):
+            cmd_path = self.command.path.render(project=self.project)
+
+            wrap(str(cmd_path), self.project)
+            return self.command.__call__(*args, project=self.project)
+
+    def __repr__(self) -> str:
+        cmd_path = self.command.path.render(project=self.project)
+
+        return f"ProjectCommand({self.project.name}, {cmd_path})"
+
+    def __str__(self) -> str:
+        return str(self.command)
+
+
+WorkloadIndex = tp.MutableMapping[WorkloadSet, tp.List[Command]]
+
+
+def unwrap(
+    index: WorkloadIndex, project: 'benchbuild.project.Project'
```

Comment:
Why do we need to pass a index here? Can unwrap be used on another index?
Normally, `project.workloads` returns the index, which could be called inside unwrap, so we could only pass project and not both.